core/camel-core-processor/src/main/java/org/apache/camel/processor/resequencer/ResequencerEngine.java

```diff
@@ -110,6 +120,30 @@ public synchronized int size() {
         return sequence.size();
     }
 
+    public void waitUntil(Predicate<Sequence<?>> pred) throws InterruptedException {
+        CountDownLatch latch;
+        synchronized (this) {
```

Comment:
_nitpick_

I don't know how exactly this class is used so this may no be an issue, but it is usually a good idea not to synchronize on the instance as you may get some side effects (got bitten in the past, so I noticed it), like the following code would end up in a deadlock:

```java
 var re = enw ResequencerEngine<>()
 syncrronized(re) {
     re.waitUntil(s -> { ... })
 }
 ```

Reply:
You're right.  I've added some comments.  However, I'm not sure there's a work around.  The conditions need to be evaluated while holding a lock on the resequencer engine, so if that lock is held by another thread, it WILL deadlock for sure.

Reply:
The conditions are evaluated when `insert()` or `deliverNext()` is called.  Those are synchronised methods.  We do synchronize inside `evaluateConditions`, but this is not really required, as the caller already holds the lock (it's more to be coherent with the `waitUntil`).  So the whole thing already deadlocks if the lock is held forever by another thread.  This won't change the behaviour.

Reply:
I don't know if applicable but a common patter is to have a _lock_ instance and lock on that

```java
private final Object lock = new Object()

public void waitUntil(Predicate<Sequence<?>> pred) throws InterruptedException {
    synchronized(lock) { ... }
}
```

but this is really a nitpick, so as you explained the issue may be also happen somewhere else

Reply:
Yes, I agree to the general rule you outline.  However, I did not want to rewrite the whole resequenced engine synchronisation, so this PR aims to integrate with the current way.  I'd suggest to use another PR to rework the synchronisation of the `ResequencerEngine` if needed.